src/content/reference/react/react-compiler.md

```diff
@@ -0,0 +1,284 @@
+
+- **Type:** `'all_errors' | 'critical_errors' | 'none'`
+- **Default:** `'none'`
+
+```js
+const ReactCompilerConfig = {
+  panicThreshold: 'none' // Recommended for production
+};
+```
+
+**Options:**
+- `'none'` (recommended): The compiler automatically skips components and hooks that break the Rules of React, allowing the rest of your code to be optimized
+- `'critical_errors'`: Causes build failures when critical errors are found
+- `'all_errors'`: Causes build failures when any diagnostic errors are found
+
+<Note>
+**Important:** The default value `'none'` is the best choice for most applications. Using `'all_errors'` or `'critical_errors'` will cause build failures and should only be used for debugging during development. Production builds should always use `'none'`.
+</Note>
+
+#### `noEmit` {/*noemit*/}
+
+Runs analysis without generating code.
+
+- **Type:** `boolean`
+- **Default:** `false`
+
+```js
+const ReactCompilerConfig = {
+  noEmit: true // Analysis only, no code generation
+};
+```
+
+#### `logger` {/*logger*/}
+
+Provides custom logging for compiler events.
+
+- **Type:** `Logger | null`
+- **Default:** `null`
+
+For example:
+
+```ts
+const ReactCompilerConfig = {
+  logger: {
+    logEvent(filename: string | null, event: LoggerEvent) {
+      switch (event.kind) {
+        case "CompileSuccess": {
+          console.log("✨ React Compiler successfully compiled: ", filename);
+          break;
+        }
+        case "CompileError": {
+          console.warn("⚠️ React Compiler failed to compile: " + filename);
+          console.warn(filename, JSON.stringify(event, null, 2));
+        }
+        default: {
+          console.log(`React Compiler: ${filename}: ${event.kind}`);
+        }
+      }
+    }
+  }
+};
+```
+
+#### `gating` {/*gating*/}
+
+Enables conditional compilation based on an external feature flag.
+
+- **Type:** `{ source: string, importSpecifierName: string } | null`
+- **Default:** `null`
+
+```js
+const ReactCompilerConfig = {
+  gating: {
+    source: 'ReactCompilerFeatureFlag',
+    importSpecifierName: 'isCompilerEnabled'
+  }
+};
+```
+
+When configured, the compiler creates two versions of each compiled function: one optimized and one original. The optimized version is conditionally used based on the imported feature flag.
+
+**Example gating module implementation:**
+
+```js
+// ReactCompilerFeatureFlag.js
+export function isCompilerEnabled() {
+  // Your logic to determine if the compiler should be enabled
+  return process.env.REACT_COMPILER_ENABLED === 'true' ||
+    Math.random() < 0.5; // A/B testing example
+}
+```
+
+With this configuration, compiled components will conditionally use optimizations:
+
+```js
+// Before compilation
+function MyComponent(props) {
+  return <div>{props.text}</div>;
+}
+
+// After compilation (simplified)
+import { isCompilerEnabled } from 'ReactCompilerFeatureFlag';
+import { c as _c } from 'react/compiler-runtime';
+
+const MyComponent = isCompilerEnabled()
```

Comment:
If that's implemented at a component level and doesn't utilize dynamic imports, can we expect a big increase in bundle size?

As effectively, we are importing double the amount of code per optimized component?

I know it won't be exactly a 2x increase, as we aren't optimizing node_modules nor normal functions, but I wonder if this would negatively impact production in a significant way.

Reply:
Yeah the number of lines per function does get doubled, but the bundle size increase isn't 1:1 with the number of lines since it depends on the compression algorithm. For us, we found that a React Native (Android) project saw an overall increase of ~10% in apk size.

There's some tradeoff math here to do, while bundle size may increase in the gating mode, you will likely see performance benefits in the `test` case that tradeoff against the bundle size. You'll need to run an experiment to really be sure. 

Of course the gating mode is optional too. The Incremental Adoption page covers a bit more about other strategies.